src/ParseFasm.cpp

```
@@ -0,0 +1,524 @@
+                std::ostringstream data;
+                data << Str(TAG('f', feature), context->FEATURE()->getText())
+                     << GET(featureAddress) << GET(value);
+                return withHeader(TAG('s', set_feature), data.str());
+        }
+
+        /// The bracketed address, where the second number is optional.
+        /// Tag: address (:)
+        virtual Any visitFeatureAddress(
+            FasmParser::FeatureAddressContext* context) override {
+                std::ostringstream data;
+                data << Num(std::stoul(context->INT(0)->getText()));
+
+                if (context->INT(1)) {
+                        data << Num(std::stoul(context->INT(1)->getText()));
+                }
+                return withHeader(TAG(':', address), data.str());
+        }
+
+        /// A Verilog style number. It can be "plain" (no leading size and
+        /// base), or hex (h), binary (b), decimal (d), or octal (o). Tag: bit
+        /// width (')
+        virtual Any visitVerilogValue(
+            FasmParser::VerilogValueContext* context) override {
+                std::ostringstream data;
+                if (context->verilogDigits()) {
+                        if (context->INT()) {
+                                data << Num(
+                                    TAG('\'', width),
+                                    std::stoi(context->INT()->getText()));
+                        }
+                        data << visit(context->verilogDigits())
+                                    .as<std::string>();
+                }
+                return data.str();
+        }
+
+        /// A "plain" decimal value.
+        /// Tag: plain (p)
+        virtual Any visitPlainDecimal(
+            FasmParser::PlainDecimalContext* context) override {
+                std::ostringstream data;
+                data << Num(TAG('p', plain),
+                            std::stol(context->INT()->getText()));
```

What does `std::stol` do on overflow?

---

It throws a `std::out_of_range` exception, so I should probably catch all exceptions in the try/catch blocks.

---

Shouldn't we handle that locally?  E.g. `line #x had plain integer constant greater than ..., which is unsupported`?

---

Done.

---

Just having all parsing errors result in a "failed to parse" isn't good enough.  The previous implementation would provide enough context (e.g. line number) and a snippet of the failing line.  Given that this is a line oriented format, providing the failing line number and reason should be sufficient.

---

Fixed.

---

Parsing errors now report line number, position, and reason.